Add Blocks API and add additional Get document descriptions. (#7836)
opensearch-project/documentation-website: _api-reference/document-apis/get-documents.md

**Introduced 1.0**
{: .label .label-purple }

After adding a JSON document to your index, you can use the Get Document API operation to retrieve the document's information and data.

## Path and HTTP methods

Use the GET method to retrieve a document and its source or stored fields from a particular index. Use the HEAD method to verify that a document exists. 

```
GET <index>/_doc/<_id>
HEAD <index>/_doc/<_id>
```

Use `_source` to retrieve the document source or verify that it exists.

```
GET <index>/_source/<_id>
HEAD <index>/_source/<_id>
```

## Query parameters

> All get document URL parameters are optional.

Reviewer: ```suggestion
All query parameters are optional.
```

## Source filtering

By default, the Get Document API returns the entire contents of the `_source` field for the requested document. However, you can choose to exclude the `_source` field from the response by using the `_source` URL parameter and setting it to false, as shown in the following example:

```json
GET test-index/_doc/0?_source=false
```

### `source` includes and excludes

If you only need specific fields from the `_source`, use the `_source_includes` or `_source_excludes` parameters to include or exclude particular fields, respectively. This can be beneficial for large documents, as retrieving only the required fields can reduce network overhead. Both parameters accept a comma-separated list of fields or wildcard expressions, as shown in the following example:

```

> GET test-index/_doc/0?_source_includes=*.id&_source_excludes=entities

Reviewer 2: Can we check if this is this a correct example ? 

if '_source_includes' only contains '*.id' , implicitly we are excluding anything else including 'entities' ?

### Distributed

The get operation gets hashed into a specific shard id. It then gets redirected to one of the replicas within that shard id and returns the result. The replicas are the primary shard and its replicas within that shard id group. This means that the more replicas we have, the better GET scaling we will have.

### Versioning support

You can use the version parameter to retrieve the document only if its current version is equal to the specified one.

Internally, OpenSearch has marked the old document as deleted and added an entirely new document. The old version of the document doesn't disappear immediately, although you won't be able to access it. OpenSearch cleans up deleted documents in the background as you continue to index more data.

## Path and HTTP methods

Use the GET method to retrieve a document and its source or stored fields from a particular index. Use the HEAD method to verify that a document exists. 

```
GET <index>/_doc/<_id>
HEAD <index>/_doc/<_id>
```

> Use `_source` to retrieve just the document source or verify that it exists.

Author: ```suggestion
Use `_source` to retrieve the document source or verify that it exists.
```

version | Integer | The version of the document to return, which must match the current version of the document.
version_type | Enum | Retrieves a specifically typed document. Available options are `external` (retrieve the document if the specified version number is greater than the document's current version) and `external_gte` (retrieve the document if the specified version number is greater than or equal to the document's current version). For example, to retrieve version 3 of a document, use `/_doc/1?version=3&version_type=external`.

> ### Realtime

Reviewer: Or if you don't want to use the param name, just Real time?

```json
GET test-index/_doc/0?_source=false
```

#### `source` includes and excludes

If you only need specific fields from the `_source`, use the `_source_includes` or `_source_excludes` parameters to include or exclude particular fields, respectively. This can be beneficial for large documents, as retrieving only the required fields can reduce network overhead. Both parameters accept a comma-separated list of fields or wildcard expressions, as shown in the following example where any `_source` that contains `*.play` is included in the response, but excludes sources with the field `entities`:

```
GET test-index/_doc/0?_source_includes=*.play&_source_excludes=entities
```

#### Shorter notation

If only want to include certain fields and don't need to exclude any, you can use a shorter notation by specifying the desired fields directly in the `_source` parameter:

```json
GET test-index/_doc/0?_source=*.id
```

### Routing

> When indexing documents in OpenSearch, specify a `routing` value to control the shard assignment for those documents. If routing was used during indexing, you must provide the same routing value when retrieving the document using the Get Document API, as shown in the following example:

Reviewer: ```suggestion
When indexing documents in OpenSearch, you can specify a `routing` value to control the shard assignment for the documents. If routing was used during indexing, you must provide the same routing value when retrieving the document using the Get Document API, as shown in the following example:
```

```json
GET test-index/_doc/1?routing=user1
```

> This request retrieves the document with the ID `1`, but it uses the routing value "user1" to determine the shard where the document is stored. If the correct routing value is not specified, the Get Document API will not be able to locate and fetch the requested document.

Reviewer: ```suggestion
This request retrieves the document with the ID `1`, but it uses the routing value "user1" to determine the shard where the document is stored. If the correct routing value is not specified, the Get Document API is not able to locate and fetch the requested document.
```

### Preference

The Get Document API allows you to control which shard replica should handle the request. By default, the operation is randomly distributed across the available shard replicas.

However, you can specify a preference to influence the replica selection. The preference can be set to one of the following values:

> - `_local`: The operation will try to execute on a locally allocated shard replica, if possible. This can improve performance by reducing network overhead.

Reviewer: ```suggestion
- `_local`: The operation tries to execute on a locally allocated shard replica, if possible. This can improve performance by reducing network overhead.
```

> - Custom (string) value: Specifying a custom string value ensures that requests with the same value will be routed to the same set of shards. This consistency can be beneficial when dealing with shards in different refresh states, as it prevents "jumping values" that may occur when hitting shards with varying data visibility. A common practice is to use a web session ID or a user name as the custom value.

Reviewer: ```suggestion
- Custom (string) value: Specifying a custom string value ensures that requests with the same value are routed to the same set of shards. This consistency can be beneficial when dealing with shards in different refresh states, as it prevents "jumping values" that may occur when hitting shards with varying data visibility. A common practice is to use a web session ID or a user name as the custom value.
```

### Distributed

When running the Get Document API, OpenSearch first calculates a hash value based on the document ID, which determines the specific shard ID where the document resides. The operation is then redirected to one of the replicas (including the primary shard and its replica shards) within that shard ID group, and the result is returned from that replica.

Having more replicas for a shard improves the scalability and performance of GET operations, as the load can be distributed across multiple replica shards. This means that the more replicas you have, the better scaling and throughput you can achieve for Get Document API requests.

### Versioning support

Use the `version` parameter to retrieve a document only if its current version matches the specified version number. This can be useful for ensuring data consistency and preventing conflicts when working with versioned documents.

> Internally, when a document is updated in OpenSearch, the old version is marked as deleted, and a new version of the document is added. However, the old version doesn't immediately disappear from the system. While you won't be able to access the old version through the Get Document API, OpenSearch handles the cleanup of deleted document versions in the background as you continue indexing new data.

Reviewer: ```suggestion
Internally, when a document is updated in OpenSearch, the original version is marked as deleted, and a new version of the document is added. However, the original version doesn't immediately disappear from the system. While you won't be able to access it through the Get Document API, OpenSearch handles the cleanup of deleted document versions in the background as you continue indexing new data.
```

## Example request

> The following example request information about a document named `1`:

Reviewer: ```suggestion
The following example requests information about a document named `1`:
```

version | Integer | The version of the document to return, which must match the current version of the document.
version_type | Enum | Retrieves a specifically typed document. Available options are `external` (retrieve the document if the specified version number is greater than the document's current version) and `external_gte` (retrieve the document if the specified version number is greater than or equal to the document's current version). For example, to retrieve version 3 of a document, use `/_doc/1?version=3&version_type=external`.

### Real time

The Get Document API in OpenSearch operates in real time by default, which means that it retrieves the latest version of the document regardless of the index's refresh rate, or the rate at which new data becomes searchable. However, if you request stored fields (using the `stored_fields` parameter) for a document that has been updated but not yet refreshed, the Get Document API parses and analyzes the document's source to extract those stored fields. 

To disable the real-time behavior and retrieve the document based on the last refreshed state of the index, set the `realtime` parameter to `false`.

### Source filtering

> By default, the Get Document API returns the entire contents of the `_source` field for the requested document. However, you can choose to exclude the `_source` field from the response by using the `_source` URL parameter and setting it to false, as shown in the following example:

Author: ```suggestion
By default, the Get Document API returns the entire contents of the `_source` field for the requested document. However, you can choose to exclude the `_source` field from the response by using the `_source` URL parameter and setting it to `false`, as shown in the following example:
```

version | Integer | The version of the document to return, which must match the current version of the document.
version_type | Enum | Retrieves a specifically typed document. Available options are `external` (retrieve the document if the specified version number is greater than the document's current version) and `external_gte` (retrieve the document if the specified version number is greater than or equal to the document's current version). For example, to retrieve version 3 of a document, use `/_doc/1?version=3&version_type=external`.

> ### Realtime

Reviewer 2: Why is the information repeated here - shouldn't this be in format similar to above older query parameters ?

Author: @bharath-techie: This is the same information. It actually belongs in this section sense it pertains more to the Request options. Deleted the other section.
